include/execution/team_thread_md_range_iterator.hpp

```diff
@@ -0,0 +1,126 @@
+                                            ///< @c Kokkos::TeamThreadRange
+  using index_type = specfem::datatype::impl::RegisterArray<
+      int, Kokkos::extents<std::size_t, rank>,
+      Kokkos::layout_left>; ///< Underlying index type. This
+                            ///< index will be passed to the
+                            ///< closure when calling @ref
+                            ///< kokkos::parallel_for with
+                            ///< this iterator.
+
+  using execution_space =
+      typename base_type::execution_space; ///< Execution space type.
+
+  /**
+   * @brief Convert linear index to 2D / 3D coordinates
+   *
+   * @param i Linear index
+   * @return index_type 2D / 3D coordinates
+   */
+  template <
+      typename T = TeamMemberType,
+      std::enable_if_t<std::is_same_v<typename T::execution_space::memory_space,
+                                      Kokkos::HostSpace>,
+                       int> = 0>
+  KOKKOS_INLINE_FUNCTION const index_type operator()(const int &i) const {
+    index_type result;
+    result(0) = i % extents_[0];
+    int temp = i / extents_[0];
+
+    for (int r = 1; r < rank - 1; ++r) {
+      result(r) = temp % extents_[r];
+      temp /= extents_[r];
+    }
+
+    if constexpr (rank > 1) {
+      result(rank - 1) = temp;
+    }
+    return result;
+  }
+
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
+
+private:
+#ifdef KOKKOS_ENABLE_CUDA
+  using device_memory_space = Kokkos::CudaSpace;
+#elif defined(KOKKOS_ENABLE_HIP)
+  using device_memory_space = Kokkos::HIPSpace;
+#endif
+public:
+  template <
+      typename T = TeamMemberType,
+      std::enable_if_t<std::is_same_v<typename T::execution_space::memory_space,
+                                      device_memory_space>,
+                       int> = 0>
+  KOKKOS_INLINE_FUNCTION const index_type operator()(const int &i) const {
+    index_type result;
+    result(rank - 1) = i % extents_[rank - 1];
+    int temp = i / extents_[rank - 1];
+
+    for (int r = rank - 2; r > 0; --r) {
+      result(r) = temp % extents_[r];
+      temp /= extents_[r];
+    }
+
+    if constexpr (rank > 1) {
+      result(0) = temp;
+    }
+    return result;
+  }
+#endif
+
+  template <typename... Indices>
+  KOKKOS_INLINE_FUNCTION constexpr
+  TeamThreadMDRangeIterator(const TeamMemberType &team, Indices... indices)
+      : base_type(team,
+                  [&]() {
+                    size_t product = 1;
+                    ((product *= indices), ...);
+                    return product;
+                  }()),
+        extents_{ indices... } {}
+};
+
+// User-defined deduction guide
+template <typename TeamMemberType>
+TeamThreadMDRangeIterator(TeamMemberType, size_t, size_t)
+    -> TeamThreadMDRangeIterator<TeamMemberType, 2>;
+
+template <typename TeamMemberType>
+TeamThreadMDRangeIterator(TeamMemberType, size_t, size_t, size_t)
+    -> TeamThreadMDRangeIterator<TeamMemberType, 3>;
```

Comment:
```suggestion
template <typename TeamMemberType, size_t... Args>
TeamThreadMDRangeIterator(TeamMemberType, Args...)
    -> TeamThreadMDRangeIterator<TeamMemberType, size_of...(Args)>
```

You might have a check the syntax

Reply:
Dont forget to update the CTAD before merge @icui